src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -0,0 +1,583 @@
+                return result;
+            }
+
+            var request = new HttpRequestMessage(HttpMethod.Get, uri);
+            log.LogInformation(string.Format(CultureInfo.InvariantCulture, "  {0} {1}.", "GET", uri));
+
+            // Read the response headers before reading the entire stream to avoid timeouts from large packages.
+            Func<Task<HttpResponseMessage>> throttleRequest = () => SendWithCredentialSupportAsync(
+                    request,
+                    HttpCompletionOption.ResponseHeadersRead,
+                    cancellationToken);
+
+            using (var response = await GetThrottled(throttleRequest, log))
+            {
+                if (ignoreNotFounds && response.StatusCode == HttpStatusCode.NotFound)
+                {
+                    log.LogInformation(string.Format(CultureInfo.InvariantCulture,
+                        "  {1} {0} {2}ms", uri, response.StatusCode.ToString(), sw.ElapsedMilliseconds.ToString()));
+                    return new HttpSourceResult();
+                }
+
+                response.EnsureSuccessStatusCode();
+
+                await CreateCacheFile(result, response, cacheContext, cancellationToken);
+
+                log.LogInformation(string.Format(CultureInfo.InvariantCulture,
+                    "  {1} {0} {2}ms", uri, response.StatusCode.ToString(), sw.ElapsedMilliseconds.ToString()));
+
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Wraps logging of the initial request and throttling.
+        /// This method does not use the cache.
+        /// </summary>
+        private async Task<HttpResponseMessage> GetAsync(
+            HttpRequestMessage request,
+            ILogger log,
+            CancellationToken cancellationToken)
+        {
+            log.LogInformation(string.Format(CultureInfo.InvariantCulture, "  {0} {1}.",
+                request.Method.Method.ToUpperInvariant(),
+                request.RequestUri));
+
+            // Read the response headers before reading the entire stream to avoid timeouts from large packages.
+            Func<Task<HttpResponseMessage>> throttledRequest = () => SendWithCredentialSupportAsync(
+                    request,
+                    HttpCompletionOption.ResponseHeadersRead,
+                    cancellationToken);
+
+            var response = await GetThrottled(throttledRequest, log);
+
+            return response;
+        }
+
+        private static async Task<HttpResponseMessage> GetThrottled(Func<Task<HttpResponseMessage>> request, ILogger log)
+        {
+            if (_throttle == null)
+            {
+                return await request();
+            }
+            else
+            {
+                try
+                {
+                    await _throttle.WaitAsync();
+
+                    log.LogVerbose($"Current http requests queued: {_throttle.CurrentCount + 1}");
```

Comment:
This is going to spam the logs and get mixed with other requests. Should we even log it?